Mission states management

yassine-kr opened the pull request:
fixes #16 
fixes #17 
fixes #19 
fixes #20 
fixes #21 
fixes #22 
fixes #23
fixes #5 
fixes #11

jcgarnier commented on api/src/main/java/cern/molr/mole/supervisor/MissionCommandRequest.java:
```diff
@@ -0,0 +1,34 @@
+    private String missionId;
+    private MoleExecutionCommand command;
+
+    public MissionCommandRequest() {
```
remove empty constructor

yassine-kr replied:
needed for deserialization

yassine-kr replied:
Removed by using @ JsonProperty annotation

jcgarnier commented on api/src/main/java/cern/molr/mole/supervisor/MissionCommandRequest.java:
```diff
@@ -0,0 +1,34 @@
+        return missionId;
+    }
+
+    public void setMissionId(String missionId) {
```
Do you have a good reason for exposing setters?

Make this class immutable, e.g. no setters, and final class.

yassine-kr replied:
done

jcgarnier commented on api/src/main/java/cern/molr/mole/supervisor/MoleCommandListener.java:
```diff
@@ -3,7 +3,7 @@
 /**
  * Listener for commands sent by supervisor to JVM
```
JVM -> MoleRunner

yassine-kr replied:
done

jcgarnier commented on api/src/main/java/cern/molr/mole/supervisor/MoleExecutionCommandResponse.java:
```diff
@@ -0,0 +1,9 @@
+package cern.molr.mole.supervisor;
+
+/**
+ * Generic result of a command request (not the result of the command execution), this data should be sent by JVM as
```
this isn't what is written in the sequence diagram on the white board.
On the white board, the command execution returns a Mono, that will provide the command response.
So it shouldn't be in the event flux.

What's the difference between the command request response and the command execution result?

yassine-kr replied:
The client gets a Mono when sending a command request to MolR server. MolR server gets a Mono when forwarding the command request, the supervisor handler gets a Mono when forwarding the command request to the supervisor service. This service calls the session controller which sends the command to the Mole runner and waits for a command status using its events reader. So the status is sent back to the controller as an event data, then it will be transformed to a CommandResponse.

jcgarnier commented on api/src/main/java/cern/molr/mole/supervisor/MoleExecutionCommandStatus.java:
```diff
@@ -1,26 +1,41 @@
 /**
- * Event sent by JVM which tells to supervisor the status of the last command received
- * @author yassine
+ * Event sent by JVM which tells to supervisor the status of the last received command
```
JVM -> MoleRunner
I'll stop commenting on this, but please address all the JVM to MoleRunner renaming

yassine-kr replied:
done

jcgarnier commented on api/src/main/java/cern/molr/mole/supervisor/MoleExecutionCommandStatus.java:
```diff
@@ -1,26 +1,41 @@
+ * Event sent by JVM which tells to supervisor the status of the last received command
+ * @author yassine-kr
  */
 public abstract class MoleExecutionCommandStatus implements MoleExecutionEvent{
```
ident this file, and others, the entire project basically, the spacing is not consistent

yassine-kr replied:
code formatted

yassine-kr commented on api/src/main/java/cern/molr/mole/Mole.java:
```diff
@@ -4,23 +4,27 @@
 
 /**
- * A {@link Mole} executes a given mission. Only a {@link Mole} knows how to run a mission.
+ * A {@link Mole} executes a given mission. Only a {@link Mole} knows how to control a mission.
```
yes just the refactoring went sideways

yassine-kr replied:
done

yassine-kr commented on api/src/main/java/cern/molr/mole/supervisor/JVMState.java:
```diff
@@ -0,0 +1,23 @@
+    /**
+     * Change the current state according to the last accepted command
+     */
+    void changeState();
```
The logic of changing the current state is encapsulated in an instance of MoleRunnerState (JVMState). This class will be modified when integrating the mole state.

yassine-kr replied:
The state change will be triggered by an event not a a command.

jcgarnier commented on api/src/main/java/cern/molr/mole/supervisor/address/AddressGetter.java:
```diff
@@ -0,0 +1,41 @@
+     *
+     * @param listener listener to be notified when the getter gets the address
+     */
+    void addListener(AddressGetterListener listener);
```
I'll need you to show me where this is used. I find the listener implementation quite heavy

yassine-kr replied:
The AddressGetter is an object that is able to return the address of the supervisor. This address has to be reachable from MolR server, the supervisor sends it in the registration request. The simple implementation is a spring component which waits for the initialization event of the server to notify its listeners after retrieving the host name and the port on which the server is running. There is one listener which sends the registration request to MolR server when notified.

yassine-kr commented on api/src/main/java/cern/molr/mole/supervisor/JVMState.java:
```diff
@@ -0,0 +1,23 @@
+     * @param command
+     * @throws Exception if the command is not accepted
+     */
+    void acceptCommand(MoleExecutionCommand command) throws CommandNotAcceptedException;
```
Does only accept or reject a command.

yassine-kr commented on api/src/main/java/cern/molr/mission/controller/custom/CustomRunMissionController.java:
```diff
@@ -2,20 +2,23 @@
  * @param <T>
  */
-public interface RunMissionController<T> {
+public interface CustomRunMissionController<T> {
```
done, the whole project architecture was improved.

yassine-kr commented on api/src/main/java/cern/molr/mission/service/MissionExecutionService.java:
```diff
@@ -4,22 +4,23 @@
-    
-    <I,O> CompletableFuture<StepMissionController<O>> step(String missionDefnClassName, I args);
+    /**
+     * A method which send a mission instantiation request to the MolR server
```
done

yassine-kr commented on api/src/main/java/cern/molr/mole/supervisor/JVMState.java:
```diff
@@ -0,0 +1,23 @@
+ * Represents a JVM state.
+ * @author yassine-kr
+ */
+public interface JVMState {
```
done

yassine-kr commented on api/src/main/java/cern/molr/server/StatefulMoleSupervisor.java:
```diff
@@ -1,12 +1,52 @@
 
 /**
- * It represents a supervisor which can tell whether it is idle or not
+ * It represents a supervisor which can gives its state
```
class deleted

yassine-kr commented on api/src/main/java/cern/molr/server/StatefulMoleSupervisor.java:
```diff
@@ -1,12 +1,52 @@
+        private int numMissions;
+        private int maxMissions;
+
+        public State() {
```
needed for deserialization

timart1n replied:
I think I would prefer to have a builder use the constructor with parameters.
@jcgarnier opinion?

yassine-kr replied:
Removed by using @ JsonProperty annotation

jcgarnier commented on api/src/main/java/cern/molr/type/ManuallySerializable.java:
```diff
@@ -0,0 +1,34 @@
+
+    /**
+     * Method which returns the map representing the json object.
+     * @return
```
you should write something in the "@return "

jcgarnier replied:
it's not the only place I saw this, please correct everywhere

yassine-kr replied:
done

jcgarnier commented on api/src/main/java/cern/molr/type/either/Either.java:
```diff
@@ -4,19 +4,23 @@
 
 /**
  * A simple data type for constructing a disjoint union (or sum type) of two types. 
  * Offers pattern matching on Either values & some powerful combinators to combine Either values
```
& -> and

yassine-kr replied:
done

jcgarnier commented:
The merge request went from 189 files to 254, and all my comments were lost. We should have synchronized so that we could finish the review before you push any changes...

yassine-kr commented:
I am still able to display your comments that you did on Friday and the ones you wrote today. Do you mean that you lost other comments because there were pushes between the beginning of the review and its end? In this case, I am sorry for your loss and I suggest fixing a time of the day when I can push changes or simply ask you if I can push my changes.

jcgarnier commented on client-api/src/main/java/cern/molr/client/api/ClientMissionController.java:
```diff
@@ -0,0 +1,36 @@
+
+
+    /**
+     * A method which send a command to the supervisor where the mission is being executed
```
command -> javadoc ref to MissionCommand

yassine-kr replied:
done

jcgarnier commented on client-api/src/main/java/cern/molr/client/api/MissionExecutionService.java:
```diff
@@ -0,0 +1,28 @@
+
+
+/**
+ * A service used by client to control a remote mission execution on a supervisor
```
"used by clients" or "used by the client"

yassine-kr replied:
done

jcgarnier commented on client/build.gradle:
```diff
@@ -1,5 +1,15 @@
+configurations.all {
+    /*This dependency is automatically required by slf4j, it should be excluded to avoid infinite loop between slf4j
```
is this true? I never had this problem. 
SLF4J as such doesn't even exist. Is it required by SLF4J-API? SLF4J-SDK?

yassine-kr replied:
Actually this dependency is imported by spring-boot-starter-webflux: https://mvnrepository.com/artifact/org.apache.logging.log4j/log4j-to-slf4j/usages
Not by slf4j as I said.

jcgarnier commented on client/src/main/java/cern/molr/client/SampleOperator.java:
```diff
@@ -0,0 +1,112 @@
+    private MissionExecutionService service;
+
+    public SampleOperator(MissionExecutionService service) {
+        this.service = service;
```
it should be checked against null

yassine-kr replied:
done

jcgarnier commented on client/src/main/java/cern/molr/client/SampleOperator.java:
```diff
@@ -0,0 +1,112 @@
+package cern.molr.client;
```
this package contains demo code and production code, split them in different packages accordingly

yassine-kr replied:
done

jcgarnier commented on client/src/main/java/cern/molr/client/SampleOperator.java:
```diff
@@ -0,0 +1,112 @@
+ *
+ * @author yassine
+ */
+public class SampleOperator {
```
why Sample? 
ExampleOpertor, or OperatorDemo

yassine-kr replied:
done

jcgarnier commented on client/src/main/java/cern/molr/client/SampleOperator.java:
```diff
@@ -0,0 +1,112 @@
+                System.out.println(execName + " event: " + event);
+                events.add(event);
+                endSignal.countDown();
+                if (event instanceof SessionInstantiated)
```
respect our coding conventions, described here: https://wikis.cern.ch/display/MPESC/Java
They mainly follow the recommendations: http://www.oracle.com/technetwork/java/javase/documentation/codeconventions-142311.html#431

so no if without curly brackets

jcgarnier commented on client/src/main/java/cern/molr/client/impl/MissionExecutionServiceImpl.java:
```diff
@@ -0,0 +1,117 @@
+
+        } catch (Exception e) {
+            LOGGER.error("error while trying to get client properties", e);
+            client = new MolrWebClient("localhost", 8000);
```
it basically means that the user asks for something, and something else can be given to him. He would see the thing running, but actually it's not what he wanted

yassine-kr replied:
The constructor of MissionExecutionServiceImpl does not try to connect to the server. It sets only the host and port address of the MolR server. Using the default constructor tries to get these values from a properties file. A failing of retrieving these values should stop the application.

yassine-kr replied:
Should we use a default configuration (default host and port)? Because I understood from your second comment that a default configuration should not be used.
Or it should only be used when the operator does not provide the properties file.

jcgarnier commented on client/src/main/java/cern/molr/client/impl/MissionExecutionServiceImpl.java:
```diff
@@ -0,0 +1,117 @@
+    }
+
+    public MissionExecutionServiceImpl(String host, int port) {
+        client = new MolrWebClient(host, port);
```
why no exception + error in this case?

yassine-kr replied:
Because the constructor just sets the host and port of the MolR server. A connection exception could be thrown when waiting for the client controller contained in the stream returned by the instantiate method of the service.

jcgarnier commented on client/src/main/java/cern/molr/client/impl/MissionExecutionServiceImpl.java:
```diff
@@ -0,0 +1,117 @@
+            clientSocket = new MolrWebSocketClient(host, port);
+
+        } catch (Exception e) {
+            LOGGER.error("error while trying to get client properties", e);
```
exception message should report more information to help someone understand the problems when they occur: here the host and port must be added

yassine-kr replied:
This exception will be thrown in two cases:
- unable to read the properties file (file does not exist for example) IOException; in this case host and port values are undefined.
- the port value cannot be parsed to an integer, so the port is undefined.

The exception should also be thrown when the host or port properties are not defined in the properties file.

So the exception message can't contain the host and port.

yassine-kr replied:
code updated

jcgarnier commented on client/src/main/java/cern/molr/client/impl/MissionExecutionServiceImpl.java:
```diff
@@ -0,0 +1,117 @@
+                            public Flux<MissionEvent> getFlux() {
+                                MissionEventsRequest eventsRequest = new MissionEventsRequest(missionExecutionId);
+                                return clientSocket.receiveFlux("/getFlux", MissionEvent.class, eventsRequest)
+                                        .doOnError((e) -> LOGGER.error("error while receiving events flux", e)).map(
```
error message shjould contain the missionExsecutionId, and missionDfnClassName, otherwise impossible to know which flux is broken

yassine-kr replied:
done

jcgarnier commented on client/src/main/java/cern/molr/client/impl/MissionExecutionServiceImpl.java:
```diff
@@ -0,0 +1,117 @@
+                                        command);
+                                return clientSocket.
+                                        receiveMono("/instruct", CommandResponse.class, commandRequest)
+                                        .doOnError((e) -> LOGGER.error("error while receiving a command response", e))
```
error message shjould contain the missionExsecutionId, and missionDfnClassName, and the commandRequest, in a readable way
otherwise impossible to know which flux is broken

yassine-kr replied:
done, displaying the command in a readable way depends on the MissionCommand implementation

jcgarnier commented on client/src/main/java/cern/molr/client/impl/MissionExecutionServiceImpl.java:
```diff
@@ -0,0 +1,117 @@
+                            }
+                        }).get());
+            } catch (InterruptedException | ExecutionException e) {
+                emitter.error(e.getCause());
```
better error message with enough diagnostic data

yassine-kr replied:
done

jcgarnier commented on commons/src/main/java/cern/molr/commons/commands/Start.java:
```diff
@@ -0,0 +1,10 @@
+
+import cern.molr.commons.request.MissionCommand;
+
+public class Start implements MissionCommand {
```
Do we actually need classes extending the MissionCommand?
It will be difficult to scale no? why not a simple String field in MissionCommand? or enums
What do you see different between each command implementation? between start, terminate, and all the custom commands.

yassine-kr replied:
We said that there will be commands that are specific to some moles. A mole developer can create its own commands by implementing the MissionCommand interface. Implementing an interface instead of using a String field would allow to the developer to create more complex commands than simple instructions.

yassine-kr replied:
The way of representing the exchanged data between the framework components should be discussed. There are four kinds of exchanged data:
- The requests, they are simple final classes.
- The commands, are classes implementing the MissionCommand Interface.
- The responses, they are represented as interfaces extending the TryResponse interface, for each response there are a success implementation and failure implementation.
- The events, are classes implementing the MissionEvent interface.

jcgarnier commented on commons/src/main/java/cern/molr/commons/events/MissionException.java:
```diff
@@ -0,0 +1,57 @@
+    private Throwable throwable;
+    private String message;
+
+    public MissionException() {
```
remove empty constructor

yassine-kr replied:
done

yassine-kr commented:
Changes commented in #8 : 

- used my github login
- used a properties file for the MolR server host and port in the client program
- the unit tests launch the servers using appropriate ports
- created the issue #19 
- the old code is removed
- applied formatting for limiting the line length to 120 chars
- custom packages removed
- the project contains more java doc
- created the issue #22 
- some typos fixed
- start and terminate methods deleted in MoleController class
- MoleSession -> MissionSession
- MoleSupervisor.isIdle() -> MoleSupervisor.getState()
- removed stateless supervisor interface
- split dependencies in their respective modules
- updated some var names
- formatting the code
- field name "gateway" -> "service" in ServerRestController class
- variable name "flux" -> "executionEventFlux" in ServerRestExecutionService class
- used CountDownLatch instead of Thread.sleep

I improved the websocket session handlers code in #24: I created a new class DataExchangeBuilder which builds a publisher that reads the received value then generates a flux. This builder is used by four classes in the project.

Project architecture updated in #24:
There are 9 modules:
- commons: it contains common api, its implementation and classes used by other modules.
- client-api: it contains only the client api.
- client: it contains an implementation of the client api, and classes for executing a sample client code.
- server-api: it contains only the server api.
- server: it contains an implementation of the server api, and classes for launching the server and handling requests.
- supervisor-api: it contains only the supervisor api.
- supervisor: it contains an implementation of the supervisor api, and classes for launching the server and handling requests.
- sample: it contains examples of missions and moles.
- test: it contains unit tests for testing the framework components.

jcgarnier commented on commons/src/main/java/cern/molr/commons/events/MissionException.java:
```diff
@@ -0,0 +1,57 @@
+        return throwable;
+    }
+
+    public void setThrowable(Throwable throwable) {
```
remove the setters, you don't need them, the exception should be built once and never be modified. It should be immutable. So no setters, and final class (like in effective java I passed to you last time)

jcgarnier replied:
or do you need setters for serialization? in this case document it

yassine-kr replied:
setters removed.
The class should not be final to allow jackson to save the class name and construct the good instance when receiving an event. The data representation of exchanged data should be discussed. If we decide to use one class for events which can represents any kind of event, we can make the class final without any problem.

jcgarnier commented on commons/src/main/java/cern/molr/commons/events/MissionFinished.java:
```diff
@@ -0,0 +1,35 @@
+ *
+ * @author yassine-kr
+ */
+public class MissionFinished implements MissionEvent {
```
final class

yassine-kr replied:
same as above

jcgarnier commented on commons/src/main/java/cern/molr/commons/events/MissionStarted.java:
```diff
@@ -0,0 +1,36 @@
+    private Object argument;
+    private String moleClassName;
+
+    public MissionStarted() {
```
remove empty constructor + make class final if possible

yassine-kr replied:
empty constructor removed, for final same as above

jcgarnier commented on commons/src/main/java/cern/molr/commons/events/SessionInstantiated.java:
```diff
@@ -0,0 +1,13 @@
+ *
+ * @author yassine-kr
+ */
+public class SessionInstantiated implements MissionEvent {
```
same than for Command, do we need one implementation per MissionEvent? or could it be more scalable

yassine-kr replied:
The data representation of exchanged data should be discussed.

jcgarnier commented on commons/src/main/java/cern/molr/commons/events/SessionInstantiated.java:
```diff
@@ -0,0 +1,13 @@
+
+/**
+ * Event sent back by the supervisor when the session is instantiated.
+ * When the JVM is created for example
```
JVM -> MoleRunner?

yassine-kr replied:
done

jcgarnier commented on commons/src/main/java/cern/molr/commons/events/SessionTerminated.java:
```diff
@@ -0,0 +1,13 @@
+
+/**
+ * Event sent back by the supervisor when the session is terminated
+ * When the JVM is destroyed for example
```
JVM -> MoleRunner ?

yassine-kr replied:
done

jcgarnier commented on commons/src/main/java/cern/molr/commons/exception/CommandNotAcceptedException.java:
```diff
@@ -0,0 +1,24 @@
+
+
+/**
+ * Thrown when the supervisor reject the execution of a command because of the current Mole runner state
```
rejects
Mole runner -> MoleRunner  == MoleRunner is actually a thing, so every time it is referenced it should be done using the same name, not sometimes MoleRunner, sometimes Mole runner, sometmies mole runner, etc. Always the same: MoleRunner

yassine-kr replied:
done

jcgarnier commented on commons/src/main/java/cern/molr/commons/exception/MissionExecutionNotAccepted.java:
```diff
@@ -0,0 +1,24 @@
+
+/**
+ * Thrown when the MolR server or a supervisor does not accept to execute a mission; for example because of the mission
+ * type or the fact that the max number of running missions is achieved
```
is achieved -> was reached

yassine-kr replied:
done

jcgarnier commented on commons/src/main/java/cern/molr/commons/exception/CommandNotAcceptedException.java:
```diff
@@ -0,0 +1,24 @@
+ */
+public class CommandNotAcceptedException extends Exception {
+
+    private static final long serialVersionUID = 195586081128114794L;
```
we usually don't use a random servialVersionUID value, but we start from 1, and increment for every change

yassine-kr replied:
done

jcgarnier commented on commons/src/main/java/cern/molr/commons/exception/MissionExecutionNotAccepted.java:
```diff
@@ -0,0 +1,24 @@
+ */
+public class MissionExecutionNotAccepted extends Exception {
+
+    private static final long serialVersionUID = 195586081128114794L;
```
same about UID starting from 1

yassine-kr replied:
done

jcgarnier commented on commons/src/main/java/cern/molr/commons/mission/Mole.java:
```diff
@@ -0,0 +1,33 @@
+/**
+ * A {@link Mole} executes a given mission. Only a {@link Mole} knows how to run a mission.
+ *
+ * @param <I>
```
javadocs missing

yassine-kr replied:
done

jcgarnier commented on commons/src/main/java/cern/molr/commons/mission/Mole.java:
```diff
@@ -0,0 +1,33 @@
+     */
+    void verify(Class<?> classType) throws IncompatibleMissionException;
+
+    O run(Mission mission, I args) throws MissionExecutionException;
```
missing javadoc

yassine-kr replied:
done

jcgarnier commented on commons/src/main/java/cern/molr/commons/request/MissionCommand.java:
```diff
@@ -0,0 +1,10 @@
+package cern.molr.commons.request;
+
+/**
+ * Generic command send by client to MolR server then forwarded to supervisor
```
sent

yassine-kr replied:
done

jcgarnier commented on commons/src/main/java/cern/molr/commons/request/MissionCommandRequest.java:
```diff
@@ -0,0 +1,35 @@
+        return missionId;
+    }
+
+    public void setMissionId(String missionId) {
```
make this class immutable by removing the setters and making the class final

yassine-kr replied:
done

jcgarnier commented on commons/src/main/java/cern/molr/commons/request/MissionCommandRequest.java:
```diff
@@ -0,0 +1,35 @@
+    private String missionId;
+    private MissionCommand command;
+
+    public MissionCommandRequest() {
```
remove

yassine-kr replied:
done

jcgarnier commented on commons/src/main/java/cern/molr/commons/request/client/MissionEventsRequest.java:
```diff
@@ -2,15 +2,19 @@
+/**
+ * Request to get flux of events from the MolR server
+ */
+public class MissionEventsRequest {
```
same about immutability and empty constructor

yassine-kr replied:
class deleted

jcgarnier commented on commons/src/main/java/cern/molr/commons/request/client/ServerInstantiationRequest.java:
```diff
@@ -2,31 +2,42 @@
+ *
+ * @author yassine-kr
+ */
+public class ServerInstantiationRequest<I> {
```
immutability
empty constructor

yassine-kr replied:
done

jcgarnier commented on commons/src/main/java/cern/molr/commons/request/server/SupervisorInstantiationRequest.java:
```diff
@@ -2,31 +2,31 @@
+ *
+ * @author yassine-kr
+ */
+public class SupervisorInstantiationRequest<I> {
```
immutability
empty constructor

yassine-kr replied:
done

jcgarnier commented on commons/src/main/java/cern/molr/commons/request/server/SupervisorStateRequest.java:
```diff
@@ -0,0 +1,30 @@
+ *
+ * @author yassine-kr
+ */
+public class SupervisorStateRequest {
```
immutability
empty constructor

yassine-kr replied:
filter field removed.
So the empty constructor could not be avoided.
I made the class final

jcgarnier commented on commons/src/main/java/cern/molr/commons/request/supervisor/SupervisorRegisterRequest.java:
```diff
@@ -1,20 +1,20 @@
- * @author yassine
+ * @author yassine-kr
  */
 public class SupervisorRegisterRequest {
```
immutability
empty constructor

yassine-kr replied:
done

jcgarnier commented on commons/src/main/java/cern/molr/commons/request/supervisor/SupervisorUnregisterRequest.java:
```diff
@@ -3,7 +3,7 @@
- * @author yassine
+ * @author yassine-kr
  */
 public class SupervisorUnregisterRequest {
```
immutability
empty constructor

yassine-kr replied:
done

jcgarnier commented on commons/src/main/java/cern/molr/commons/response/Ack.java:
```diff
@@ -2,19 +2,20 @@
+ * @author yassine-kr
  */
-public class Ack{
+public class Ack {
```
immutability
empty constructor

timart1n replied:
Yes I pointed out that this should be done everywhere...

yassine-kr replied:
done

jcgarnier commented on commons/src/main/java/cern/molr/commons/response/InstantiationResponseBean.java:
```diff
@@ -4,19 +4,19 @@
+ *
+ * @author yassine-kr
+ */
+public class InstantiationResponseBean {
```
immutability
empty constructor

yassine-kr replied:
done

timart1n commented on commons/src/main/java/cern/molr/commons/request/MissionCommandRequest.java:
```diff
@@ -0,0 +1,35 @@
+    private String missionId;
+    private MissionCommand command;
+
+    public MissionCommandRequest() {
```
Delete this constructor

yassine-kr replied:
done

timart1n commented on commons/src/main/java/cern/molr/commons/request/MissionCommandRequest.java:
```diff
@@ -0,0 +1,35 @@
+ *
+ * @author yassine-kr
+ */
+public class MissionCommandRequest {
```
missing hash and equals method

yassine-kr replied:
done

timart1n commented on commons/src/main/java/cern/molr/commons/request/MissionCommandRequest.java:
```diff
@@ -0,0 +1,35 @@
+        return missionId;
+    }
+
+    public void setMissionId(String missionId) {
```
If we create it using the constructor we should delete these sets.

yassine-kr replied:
done

timart1n commented on commons/src/main/java/cern/molr/commons/request/client/MissionEventsRequest.java:
```diff
@@ -2,15 +2,19 @@
-    public MissionCancelRequest(){}
-    
-    public MissionCancelRequest(String missionExecutionId){
+    public MissionEventsRequest() {
```
Delete this

yassine-kr replied:
done

yassine-kr replied:
class deleted

timart1n commented on commons/src/main/java/cern/molr/commons/request/client/ServerInstantiationRequest.java:
```diff
@@ -2,31 +2,42 @@
+ */
+public class ServerInstantiationRequest<I> {
 
     private String missionDefnClassName;
```
missionName

yassine-kr replied:
done

timart1n commented on commons/src/main/java/cern/molr/commons/request/client/ServerInstantiationRequest.java:
```diff
@@ -2,31 +2,42 @@
+public class ServerInstantiationRequest<I> {
 
     private String missionDefnClassName;
     private I args;
```
missionArguments to be more verbose

yassine-kr replied:
done

timart1n commented on commons/src/main/java/cern/molr/commons/request/client/ServerInstantiationRequest.java:
```diff
@@ -2,31 +2,42 @@
-    
-    public MissionExecutionRequest(String missionDefnClassName, I args){
+
+    public ServerInstantiationRequest() {
```
delete this

yassine-kr replied:
done

timart1n commented on commons/src/main/java/cern/molr/commons/request/client/ServerInstantiationRequest.java:
```diff
@@ -2,31 +2,42 @@
+    public ServerInstantiationRequest() {
+    }
+
+    public ServerInstantiationRequest(String missionDefnClassName, I args) {
```
missionName and missionArguments

yassine-kr replied:
done

timart1n commented on commons/src/main/java/cern/molr/commons/request/client/ServerInstantiationRequest.java:
```diff
@@ -2,31 +2,42 @@
+ *
+ * @author yassine-kr
+ */
+public class ServerInstantiationRequest<I> {
```
missing hash and equals

yassine-kr replied:
done

timart1n commented on commons/src/main/java/cern/molr/commons/request/server/SupervisorInstantiationRequest.java:
```diff
@@ -2,31 +2,31 @@
     private String missionExecutionId;
-    private String moleClassName;
     private String missionContentClassName;
     private I args;
```
missionName and missionArguments

yassine-kr replied:
done

timart1n commented on commons/src/main/java/cern/molr/commons/request/server/SupervisorInstantiationRequest.java:
```diff
@@ -2,31 +2,31 @@
-    
-    public MissionExecutionRequest(String missionExecutionId, String moleClassName, String missionContentClassName, I args){
+
+    public SupervisorInstantiationRequest() {
```
delete this

yassine-kr replied:
done

timart1n commented on commons/src/main/java/cern/molr/commons/request/server/SupervisorInstantiationRequest.java:
```diff
@@ -2,31 +2,31 @@
+    public SupervisorInstantiationRequest() {
+    }
+
+    public SupervisorInstantiationRequest(String missionExecutionId,
```
same here, names

yassine-kr replied:
done

timart1n commented on commons/src/main/java/cern/molr/commons/request/server/SupervisorInstantiationRequest.java:
```diff
@@ -2,31 +2,31 @@
+ *
+ * @author yassine-kr
+ */
+public class SupervisorInstantiationRequest<I> {
```
missing hash and equals

timart1n commented on commons/src/main/java/cern/molr/commons/request/server/SupervisorStateRequest.java:
```diff
@@ -0,0 +1,30 @@
+ *
+ * @author yassine-kr
+ */
+public class SupervisorStateRequest {
```
missing hash and equals

yassine-kr replied:
done

timart1n commented on commons/src/main/java/cern/molr/commons/request/server/SupervisorStateRequest.java:
```diff
@@ -0,0 +1,30 @@
+
+    private String filter;
+
+    public SupervisorStateRequest() {
```
delete this

yassine-kr replied:
done

timart1n commented on commons/src/main/java/cern/molr/commons/request/server/SupervisorStateRequest.java:
```diff
@@ -0,0 +1,30 @@
+ */
+public class SupervisorStateRequest {
+
+    private String filter;
```
What is this filter used for?

yassine-kr replied:
I thought that it could be useful in the future.

timart1n replied:
Lets not add things that are not necessary.
Keep it simple. Later if we need it we add it.

yassine-kr replied:
done

timart1n commented on commons/src/main/java/cern/molr/commons/request/supervisor/SupervisorRegisterRequest.java:
```diff
@@ -1,20 +1,20 @@
- * @author yassine
+ * @author yassine-kr
  */
 public class SupervisorRegisterRequest {
```
Just noticed that all of the fields on these DTOs should be final, for this and all the prevous ones.

timart1n replied:
also remove sets and add hash and equals

yassine-kr replied:
done for requests classes and some other classes

timart1n commented on commons/src/main/java/cern/molr/commons/request/supervisor/SupervisorRegisterRequest.java:
```diff
@@ -1,20 +1,20 @@
     private List<String> acceptedMissions;
 
-    public SupervisorRegisterRequest(){}
+    public SupervisorRegisterRequest() {
```
delete this guy

yassine-kr replied:
done

timart1n commented on commons/src/main/java/cern/molr/commons/request/supervisor/SupervisorUnregisterRequest.java:
```diff
@@ -3,7 +3,7 @@
- * @author yassine
+ * @author yassine-kr
  */
 public class SupervisorUnregisterRequest {
```
final field, remove empty constructor and add hash and equals

yassine-kr replied:
done

jcgarnier commented on commons/src/main/java/cern/molr/commons/response/SupervisorState.java:
```diff
@@ -0,0 +1,37 @@
+
+import com.fasterxml.jackson.annotation.JsonIgnore;
+
+public class SupervisorState {
```
immutable 
empty constructor

timart1n replied:
@jcgarnier 
What do you think of trying https://immutables.github.io/.
Since we have A LOT of these types of objects.

jcgarnier commented on commons/src/main/java/cern/molr/commons/response/SupervisorStateResponse.java:
```diff
@@ -0,0 +1,44 @@
+            implements SupervisorStateResponse {
+
+        public SupervisorStateResponseSuccess() {
+            super(null);
```
could we use optionals?

timart1n replied:
In my opinion I dont see why is this even here.
I would prefer to have a generic response with a code for ok and for failure.

yassine-kr replied:
The responses representation should be discussed.

jcgarnier commented on commons/src/main/java/cern/molr/commons/response/SupervisorUnregisterResponse.java:
```diff
@@ -1,69 +1,45 @@
 
 /**
- * Interface representing response to an unregister request
+ * Response to a unregister request sent back by the MolR server
```
an unregistered

yassine-kr replied:
I think an unregistration  request

timart1n commented on commons/src/main/java/cern/molr/commons/request/supervisor/SupervisorUnregisterRequest.java:
```diff
@@ -3,7 +3,7 @@
- * @author yassine
+ * @author yassine-kr
  */
 public class SupervisorUnregisterRequest {
```
I'm also seeing a lot of DTOs which only have one field inside.
I think these should be deleted and replaced in the code by the specific type.

On the supervisor why should I do:
```java
molrWebClient.post("/unregisterSupervisor", SupervisorUnregisterRequest.class, new SupervisorUnregisterRequest(10), SupervisorUnregisterResponse.class)
```

When I can simply create a ServerSupervisorClient or something, that extends the MolrWebClient and do:
```java
client.unregisterSupervisor(10)
```

and have it reuse the post inside of the MolrClient.
This makes the client API much simpler and hides some ugly details.

This comment is also valid in other places where the web client is used.

jcgarnier commented on commons/src/main/java/cern/molr/commons/type/trye/TryResponseFailure.java:
```diff
@@ -0,0 +1,41 @@
+ * @author ?
+ * @author yassine-kr
+ */
+public class TryResponseFailure<T> extends Failure<T> implements TryResponse<T> {
```
immutable and empty constructor

yassine-kr replied:
empty constructor removed.
The class can't be final because it is extended
setters removed by using some Jackson annotations

jcgarnier commented on commons/src/main/java/cern/molr/commons/web/DataExchangeBuilder.java:
```diff
@@ -0,0 +1,171 @@
+    }
+
+    /**
+     * A function which can throws an exception
```
can throw

yassine-kr replied:
done

jcgarnier commented on commons/src/main/java/cern/molr/commons/web/MolrWebSocketClient.java:
```diff
@@ -1,54 +1,94 @@
-        });
-        */
+        /**
+         * Processor which which receives data from server
```
which which

yassine-kr replied:
done

timart1n commented on commons/src/main/java/cern/molr/commons/response/CommandResponse.java:
```diff
@@ -0,0 +1,64 @@
+
+
+        @Override
+        public Map<String, String> getJsonMap() {
```
This is a no no for me. We have a json serializer/desserializer implementation for a reason. We should never do this ourselves.

yassine-kr replied:
done

timart1n commented on commons/src/main/java/cern/molr/commons/response/InstantiationResponse.java:
```diff
@@ -0,0 +1,49 @@
+ *
+ * @author yassine-kr
+ */
+public interface InstantiationResponse extends TryResponse<InstantiationResponseBean> {
```
I don't know but Im still not convined by this TryResponse thing. I believe this comes from nachis code? (he had something similar with the either).
Why not simply have a response class with a string or an int with the return code?

yassine-kr replied:
Should be discussed

timart1n commented on commons/src/main/java/cern/molr/commons/response/InstantiationResponse.java:
```diff
@@ -0,0 +1,49 @@
+    class InstantiationResponseSuccess extends TryResponseSuccess<InstantiationResponseBean>
+            implements InstantiationResponse {
+
+        public InstantiationResponseSuccess() {
```
Again remove this

yassine-kr replied:
done

timart1n commented on commons/src/main/java/cern/molr/commons/response/ManuallySerializable.java:
```diff
@@ -0,0 +1,39 @@
+ *
+ * @author yassine-kr
+ */
+public interface ManuallySerializable {
```
I don't like this and in my opinion does not make sense.
What if the manual serialization fails? Are we gonna have a magic serialization which is used when the manual serialization fails, which is used when the normal serialization fails? See my point?
The serialization either passes or we have a problem and we should solve it.

yassine-kr replied:
interface deleted

timart1n commented on commons/src/main/java/cern/molr/commons/response/ManuallySerializable.java:
```diff
@@ -0,0 +1,39 @@
+     *
+     * @return
+     */
+    static String serializeArray(ManuallySerializable object) {
```
Seriously this looks really bad...

yassine-kr replied:
done

timart1n commented on commons/src/main/java/cern/molr/commons/response/SupervisorRegisterResponse.java:
```diff
@@ -1,68 +1,45 @@
  */
-@JsonDeserialize(using = SupervisorRegisterResponse.SupervisorRegisterResponseDeserializer.class)
-public interface SupervisorRegisterResponse extends TryResponse<SupervisorRegisterResponseBean>{
+public interface SupervisorRegisterResponse extends TryResponse<SupervisorRegisterResponseBean> {
```
Interface which only have 2 clases defined inside? Hummm something is wrong here.

yassine-kr replied:
I used the same structure used by @nachivpn

timart1n commented on commons/src/main/java/cern/molr/commons/response/SupervisorRegisterResponseBean.java:
```diff
@@ -1,22 +1,17 @@
- * @author yassine
+ * @author yassine-kr
  */
 public class SupervisorRegisterResponseBean {
```
Final field
remove empty constructor 
remove bean from name (everywhere)

yassine-kr replied:
final field: done
empty constructor removed
there is an interface which is named SupervisorRegisterResponse.
The same for InstantiationResponseBean

timart1n commented on commons/src/main/java/cern/molr/commons/response/SupervisorUnregisterResponse.java:
```diff
@@ -1,69 +1,45 @@
  */
-@JsonDeserialize(using = SupervisorUnregisterResponse.SupervisorUnregisterResponseDeserializer.class)
-public interface SupervisorUnregisterResponse extends TryResponse<Ack>{
+public interface SupervisorUnregisterResponse extends TryResponse<Ack> {
```
Again we have an interface which has 2 classes inside...

yassine-kr replied:
@nachivpn code

timart1n commented on commons/src/main/java/cern/molr/commons/type/trye/Failure.java:
```diff
@@ -2,28 +2,26 @@
-   
+public class Failure<T> extends Left<Throwable, T> implements Try<T> {
+
     public Failure() {
```
Remove this

yassine-kr replied:
What should I remove exactly?

timart1n replied:
Empty constructor.

yassine-kr replied:
done

timart1n commented on commons/src/main/java/cern/molr/commons/type/trye/Success.java:
```diff
@@ -2,31 +2,28 @@
+ * @author yassine-kr
  */
-public class Success<T> extends Right<Throwable, T> implements Try<T>{
+public class Success<T> extends Right<Throwable, T> implements Try<T> {
```
I think we should defenitly take a look into this sooner or later...
Just to return a successcful answer, which should simply be returning the correct value which would be serialized by the serializer we have to use a Success<T> that extends Right<Throwable, T> that implements Either<Throwable, T>.

Also of note that right Success<T> also implements Try<T> which extends Either<throwable, T>.

And this brings zero benefit as the conternder solution would simply to return T...
I seriously don't get this and HIGHLY advise in removing all of it...

yassine-kr replied:
Yes it is a bit complex without a real benefit.

timart1n commented on commons/src/main/java/cern/molr/commons/type/trye/Success.java:
```diff
@@ -2,31 +2,28 @@
-    /**
-     * @param r
-     */
     public Success() {
```
Remove this

yassine-kr replied:
done

timart1n commented on commons/src/main/java/cern/molr/commons/web/DataExchangeBuilder.java:
```diff
@@ -0,0 +1,171 @@
+     *
+     * @return this builder to chain other methods
+     */
+    public DataExchangeBuilder<Input, Output> setPreInput(Flux<String> preInput) {
```
This class is too complex... Please extract meaningful methods so we can understand what is actually going on here.

yassine-kr replied:
As I removed the manual serialization, the code is now more simple

timart1n commented on commons/src/main/java/cern/molr/commons/web/MolrWebClient.java:
```diff
@@ -4,30 +4,60 @@
-        return CompletableFuture.supplyAsync(
+    public <I, O> CompletableFuture<O> post(String uri, Class<I> requestClass, I request, Class<O> responseClass) {
+
+        ExecutorService executorService = Executors.newSingleThreadExecutor();
```
Create a configurable executor in construction time.

yassine-kr replied:
executor service deleted

timart1n commented on commons/src/main/java/cern/molr/commons/web/MolrWebClient.java:
```diff
@@ -4,30 +4,60 @@
 
-    public <I,O> CompletableFuture<O> post(String uri, Class<I> requestClass, I request, Class<O> responseClass) {
-        return CompletableFuture.supplyAsync(
+    public <I, O> CompletableFuture<O> post(String uri, Class<I> requestClass, I request, Class<O> responseClass) {
```
I though we had agreed in using reactive only instead of mixing with futures at the end.

yassine-kr replied:
done

timart1n commented on commons/src/main/java/cern/molr/commons/web/MolrWebClient.java:
```diff
@@ -4,30 +4,60 @@
+                    clientDefaultCodecsConfigurer.defaultCodecs()
+                            .jackson2JsonDecoder(new Jackson2JsonDecoder(mapper, MediaType.APPLICATION_JSON));
+                }).build();
+        this.client = WebClient.builder().baseUrl("http://" + host + ":" + port).exchangeStrategies(strategies).build();
```
http must be part of the host as we might decide to go for https in the future

yassine-kr replied:
done

timart1n commented on commons/src/main/java/cern/molr/commons/web/MolrWebClient.java:
```diff
@@ -4,30 +4,60 @@
+                        .accept(MediaType.APPLICATION_JSON)
+                        .body(BodyInserters.fromPublisher(Mono.just(request), requestClass)).exchange()
+                        .flatMapMany(value -> value.bodyToMono(responseClass))
+                        .doOnError(e -> {
```
Small name

yassine-kr replied:
Do you mean the class name MolrWebClient or the parameter names requestClass and responseClass ?

timart1n replied:
I mean e. Use error or something similar.

yassine-kr replied:
done

timart1n commented on commons/src/main/java/cern/molr/commons/web/MolrWebSocketClient.java:
```diff
@@ -1,54 +1,94 @@
+     * @return try elements, if the serialization of an element throws an exception, this exception is pushed to flux
      */
-    public <T> Mono<Flux<T>> receiveFlux(Class<T> type){
+    public <I, T> Flux<Try<T>> receiveFlux(String path, Class<T> responseType, I request) {
```
reactive api

yassine-kr replied:
done

yassine-kr commented on .gitignore:
```diff
@@ -9,6 +9,9 @@
 *.war
 *.ear
 
+!gradle-wrapper.jar
```
I thought that travis needs this file to build the project. I did the same as this repository:
https://github.com/streamingpool/streamingpool-core/tree/master/gradle/wrapper

timart1n replied:
Travis needs it but if you remove this line it will be there

yassine-kr replied:
It will be there if the .gitignore does not ignore it
```
# Package Files #
*.jar
*.war
*.ear
```
So I added an exception to accept gradle-wrapper.jar file

timart1n replied:
Ok leave it, we can change it later. I prefer to be more specific in my gitignores instead of using *.jar

yassine-kr commented on api/src/main/java/cern/molr/commons/AnnotatedMissionMaterializer.java:
```diff
@@ -38,7 +35,8 @@ public Mission materialize(Class<?> classType) throws MissionMaterializationExce
         RunWithMole moleAnnotation = classType.getAnnotation(RunWithMole.class);
         if (null == moleAnnotation) {
-            throw new MissionMaterializationException(new IllegalArgumentException(String.format("Class type [%s] is not annotated with RunWithMole", classType)));
+            throw new MissionMaterializationException(new IllegalArgumentException(String.format("Class type [%s] is " +
```
Where is the weird thing?

timart1n replied:
When I was revieweing it had a weitrd formatting but it could be just my browser windows misbehaving. If its ok then there is nothing to do.

yassine-kr commented on build.gradle:
```diff
@@ -2,71 +2,96 @@ allprojects {
     apply plugin: "idea"
```
done

yassine-kr commented on build.gradle:
```diff
@@ -2,71 +2,96 @@ allprojects {
+}
+
+project(":client") {
+    dependencies {
```
done

yassine-kr commented on client-api/src/main/java/cern/molr/client/api/MissionDeploymentService.java:
```diff
@@ -2,17 +2,17 @@
 
-    public void deploy(Mission m);
-    
+    void deploy(String missionClassName);
```
done

yassine-kr commented on client-api/src/main/java/cern/molr/client/api/MissionExecutionService.java:
```diff
@@ -0,0 +1,28 @@
+     * @param args             the mission arguments, can be null if the mission does not need any argument
+     * @param <I>              the argument type
+     *
+     * @return A future mission controller
```
done

yassine-kr commented on client/src/main/java/cern/molr/client/SampleOperator.java:
```diff
@@ -0,0 +1,112 @@
+package cern.molr.client;
```
done

yassine-kr commented on client/src/main/java/cern/molr/client/impl/MissionExecutionServiceImpl.java:
```diff
@@ -0,0 +1,117 @@
+        } catch (Exception e) {
+            LOGGER.error("error while trying to get client properties", e);
+            client = new MolrWebClient("localhost", 8000);
+            clientSocket = new MolrWebSocketClient("localhost", 8000);
```
When should we use the default configuration?

yassine-kr commented on client/src/main/java/cern/molr/client/impl/MissionExecutionServiceImpl.java:
```diff
@@ -0,0 +1,117 @@
+            client = new MolrWebClient(host, port);
+            clientSocket = new MolrWebSocketClient(host, port);
+
+        } catch (Exception e) {
```
done

yassine-kr commented on client/src/main/java/cern/molr/client/impl/MissionExecutionServiceImpl.java:
```diff
@@ -0,0 +1,117 @@
+    }
+
+    @Override
+    public <I> Mono<ClientMissionController> instantiate(String missionDefnClassName, I args) {
```
done

yassine-kr commented on client/src/main/java/cern/molr/client/impl/MissionExecutionServiceImpl.java:
```diff
@@ -0,0 +1,117 @@
+
+        ExecutorService executorService = Executors.newSingleThreadExecutor();
+
+        ServerInstantiationRequest<I> execRequest = new ServerInstantiationRequest<>(missionDefnClassName, args);
```
done

yassine-kr commented on client/src/main/java/cern/molr/client/impl/MissionExecutionServiceImpl.java:
```diff
@@ -0,0 +1,117 @@
+                            public Flux<MissionEvent> getFlux() {
+                                MissionEventsRequest eventsRequest = new MissionEventsRequest(missionExecutionId);
+                                return clientSocket.receiveFlux("/getFlux", MissionEvent.class, eventsRequest)
+                                        .doOnError((e) -> LOGGER.error("error while receiving events flux", e)).map(
```
done

yassine-kr commented on client/src/main/java/cern/molr/client/impl/MissionExecutionServiceImpl.java:
```diff
@@ -0,0 +1,117 @@
+                                        .doOnError((e) -> LOGGER.error("error while receiving events flux", e)).map(
+                                                (tryElement)
+                                                        -> tryElement
+                                                        .match(MissionException::new, Function.identity()));
```
A tryElement is an instance implementing the Either interface, the match method allows to apply a transformation to the instance (can for example return another instance of a new type). It has two parameters, the first one is the function to apply if the instance is a left value, and the second one is a function to apply if the instance is a right value. The two function must have the same return type.

yassine-kr commented on commons/src/main/java/cern/molr/commons/events/MissionFinished.java:
```diff
@@ -0,0 +1,35 @@
+    private Object result;
+    private String moleClassName;
+
+    public MissionFinished() {
```
Yes in the class MoleRunner when the mission is finished

yassine-kr replied:
empty constructor removed

yassine-kr commented on commons/src/main/java/cern/molr/commons/events/MissionFinished.java:
```diff
@@ -0,0 +1,35 @@
+    public MissionFinished() {
+    }
+
+    public MissionFinished(String missionClassName, Object result, String moleClassName) {
```
done

yassine-kr commented on commons/src/main/java/cern/molr/commons/mission/AnnotatedMissionMaterializer.java:
```diff
@@ -4,32 +4,38 @@
  * @author tiagomr
+ * @author yassine-kr
  */
 public class AnnotatedMissionMaterializer implements MissionMaterializer {
```
It was actually used by the deployment service (not yet implemented code). So I made it in commons module because I thought it could be needed by the client code or the server code too when implementing the infrastructure layer.

yassine-kr commented on client/src/main/java/cern/molr/client/impl/MissionExecutionServiceImpl.java:
```diff
@@ -0,0 +1,117 @@
+        ServerInstantiationRequest<I> execRequest = new ServerInstantiationRequest<>(missionDefnClassName, args);
+        Mono<ClientMissionController> mono = Mono.<ClientMissionController>create((emitter) -> {
+            try {
+                emitter.success(client.post("/instantiate", ServerInstantiationRequest.class, execRequest,
```
done

yassine-kr commented on api/src/main/java/cern/molr/server/StatefulMoleSupervisor.java:
```diff
@@ -1,12 +1,52 @@
+            return numMissions;
+        }
+
+        public void setNumMissions(int numMissions) {
```
done

yassine-kr commented on client-api/src/main/java/cern/molr/client/api/ClientMissionController.java:
```diff
@@ -0,0 +1,36 @@
+     *
+     * @return
+     */
+    Flux<MissionEvent> getFlux();
```
done

yassine-kr commented on client-api/src/main/java/cern/molr/client/api/ClientMissionController.java:
```diff
@@ -0,0 +1,36 @@
+     *
+     * @return a future response which says whether the command was accepted or not
+     */
+    Mono<CommandResponse> instruct(MissionCommand command);
```
done

yassine-kr commented on client-api/src/main/java/cern/molr/client/api/MissionExecutionService.java:
```diff
@@ -0,0 +1,28 @@
+     *
+     * @return A future mission controller
+     */
+    <I> Mono<ClientMissionController> instantiate(String missionClassName, I args);
```
done

yassine-kr commented on client/src/main/java/cern/molr/client/impl/MissionExecutionServiceImpl.java:
```diff
@@ -0,0 +1,117 @@
+    @Override
+    public <I> Mono<ClientMissionController> instantiate(String missionDefnClassName, I args) {
+
+        ExecutorService executorService = Executors.newSingleThreadExecutor();
```
the executor service is not used anymore in MissionServiceImpl

yassine-kr commented on client/src/main/java/cern/molr/client/impl/MissionExecutionServiceImpl.java:
```diff
@@ -0,0 +1,117 @@
+                            @Override
+                            public Flux<MissionEvent> getFlux() {
+                                MissionEventsRequest eventsRequest = new MissionEventsRequest(missionExecutionId);
+                                return clientSocket.receiveFlux("/getFlux", MissionEvent.class, eventsRequest)
```
done

yassine-kr commented on client/src/main/java/cern/molr/client/impl/MissionExecutionServiceImpl.java:
```diff
@@ -0,0 +1,117 @@
+                                MissionCommandRequest commandRequest = new MissionCommandRequest(missionExecutionId,
+                                        command);
+                                return clientSocket.
+                                        receiveMono("/instruct", CommandResponse.class, commandRequest)
```
done

yassine-kr commented on client/src/main/java/cern/molr/client/impl/MissionExecutionServiceImpl.java:
```diff
@@ -0,0 +1,117 @@
+                emitter.error(e.getCause());
+            }
+        }).subscribeOn(Schedulers.fromExecutorService(executorService)).doOnTerminate(() -> executorService.shutdown());
+        return mono;
```
As I added methods to the socket client allowing it to create connection session for each kind of request (instantiate, getEventsStream, and instruct), the code of the MissionExecutionServiceImpl instantiate method became more simple.

yassine-kr commented on commons/src/main/java/cern/molr/commons/events/MissionStarted.java:
```diff
@@ -0,0 +1,36 @@
+    private Object argument;
+    private String moleClassName;
+
+    public MissionStarted() {
```
empty constructor removed

yassine-kr commented on commons/src/main/java/cern/molr/commons/mission/MissionImpl.java:
```diff
@@ -43,30 +33,20 @@ public MissionImpl(String moleClassName, String missionContentClassName) {
-        this.moleClassName = moleClassName;
-    }
-
     public void setMissionDefnClassName(String missionContentClassName) {
```
done

yassine-kr commented on commons/src/main/java/cern/molr/commons/mission/MissionImpl.java:
```diff
@@ -78,7 +58,8 @@ public boolean equals(Object o) {
         MissionImpl service = (MissionImpl) o;
```
done

yassine-kr commented on commons/src/main/java/cern/molr/commons/mission/MissionMaterializer.java:
```diff
@@ -2,17 +2,17 @@
+ *
  * @author nachivpn
  */
 public interface MissionMaterializer {
```
done

yassine-kr commented on commons/src/main/java/cern/molr/commons/mission/Mole.java:
```diff
@@ -0,0 +1,33 @@
+ * @author nachivpn
+ * @author yassine-kr
+ */
+public interface Mole<I, O> {
```
done
